Mathlib/Tactic/Linter/UnusedTactic.lean

```diff
@@ -0,0 +1,213 @@
+    |>.insert ``Lean.Parser.Tactic.failIfSuccess
+    |>.insert `Mathlib.Tactic.successIfFailWithMsg
+    |>.insert `Mathlib.Tactic.failIfNoProgress
+
+/-- Is this a syntax kind that contains intentionally unused tactic subterms? -/
+def isIgnoreTacticKind (ignoreTacticKinds : NameHashSet) (k : SyntaxNodeKind) : Bool :=
+  k.components.contains `Conv ||
+  "slice".isPrefixOf k.toString ||
+  match k with
+  | .str _ "quot" => true
+  | _ => ignoreTacticKinds.contains k
+
+/--
+Adds a new syntax kind whose children will be ignored by the `unusedTactic` linter.
+This should be called from an `initialize` block.
+-/
+def addIgnoreTacticKind (kind : SyntaxNodeKind) : IO Unit :=
+  ignoreTacticKindsRef.modify (·.insert kind)
+
+variable (ignoreTacticKinds : NameHashSet) (isTacKind : SyntaxNodeKind → Bool) in
+/-- Accumulates the set of tactic syntaxes that should be evaluated at least once. -/
+@[specialize] partial def getTactics (stx : Syntax) : M Unit := do
+  if let .node _ k args := stx then
+    if !isIgnoreTacticKind ignoreTacticKinds k then
+      args.forM getTactics
+    if isTacKind k then
+      if let some r := stx.getRange? true then
+        modify fun m => m.insert r stx
+
+/-- `getNames mctx` extracts the names of all the local declarations implied by the
+`MetavarContext` `mctx`. -/
+def getNames (mctx : MetavarContext) : List Name :=
+  let lcts := mctx.decls.toList.map (MetavarDecl.lctx ∘ Prod.snd)
+  let locDecls := (lcts.map (PersistentArray.toList ∘ LocalContext.decls)).join.reduceOption
+  locDecls.map LocalDecl.userName
+
+mutual
+/-- Search for tactic executions in the info tree and remove the syntax of the tactics that
+changed something. -/
+partial def eraseUsedTacticsList (trees : PersistentArray InfoTree) : M Unit :=
+  trees.forM eraseUsedTactics
+
+/-- Search for tactic executions in the info tree and remove the syntax of the tactics that
+changed something. -/
+partial def eraseUsedTactics : InfoTree → M Unit
+  | .node i c => do
+    if let .ofTacticInfo i := i then
+      let stx := i.stx
+      let kind := stx.getKind
+      if let some r := stx.getRange? true then
+        if allowed.contains kind
+        -- if the tactic is allowed to not change the goals
+        then modify (·.erase r)
+        else
+        -- if the goals have changed
+        if i.goalsAfter != i.goalsBefore
+        then modify (·.erase r)
+        -- bespoke check for `swap_var`: the only change that it does is
+        -- in the usernames of local declarations, so we check the names before and after
+        else
+        if (kind == `Mathlib.Tactic.«tacticSwap_var__,,») &&
+                (getNames i.mctxBefore != getNames i.mctxAfter)
+        then modify (·.erase r)
+    eraseUsedTacticsList c
+  | .context _ t => eraseUsedTactics t
+  | .hole _ => pure ()
+
+end
+
+/-- Gets the value of the `linter.unusedTactic` option. -/
+def getLinterHash (o : Options) : Bool := Linter.getLinterValue linter.unusedTactic o
```

Comment:
```suggestion
def getLinterHash (o : Options) : Bool := linter.unusedTactic.getLinterValue o
```
no?

Reply:
Thank you very much for the reviews, Yaël!

I tried this suggestion, but it does not work: the type of `linter.unusedTactic` is `Lean.Option Bool`.  The `Linter` itself is the *following* declaration, `unusedTacticLinter`.